### ISSUE TITLE
Hyper CLI

### ISSUE THREAD

**Issue by albinekb** [zeit/hyper#2288]
As started in https://github.com/zeit/hyper/pull/282, we will in 2.0 have a way to manage hyper via the CLI.

**Comment by chabou** [comment]
I spent some time to figure out how to do this. And here are my thoughts.
Any feedback will be warmly welcomed (cc @rauchg @albinekb @leo)

**TL;DR** Embedding `hpm` is pretty obvious and straight-forward. The most difficult part is to add a launch script that is in user PATH.

### How VSCode works?
`vscode` is a good exemple because this is an Electron app that have a `code` command to install/uninstall/list plugins or launch app for a given directory/file.

#### Javascript CLI-Tool
After an amd-boostraping in [src/cli.js](https://github.com/Microsoft/vscode/blob/e29c517386fe6f3a40e2f0ff00effae4919406aa/src/cli.js) (copied to `"$VSCODE_PATH/resources/app/out/cli.js"` at build time), it executes the real CLI entry point: [src/vs/code/node/cli.ts](https://github.com/Microsoft/vscode/tree/e29c517386fe6f3a40e2f0ff00effae4919406aa/src/vs/code/node/cli.ts)

This entry point is responsible to parse args and to intercept/deal with some options like `--help`/`--version` or to launch app for a given directory/file. If option concern plugin management, it calls [src/vs/code/node/cliProcessMain.js](https://github.com/Microsoft/vscode/tree/e29c517386fe6f3a40e2f0ff00effae4919406aa/src/vs/code/node/cliProcessMain.js).

#### Shell command
How shell can execute this javascript file `"$VSCODE_PATH/resources/app/out/cli.js"`?
For every platform, there is a script that mainly resolve paths and launch Electron as node on this file.
[resources/(darwin|windows|linux)/code.(sh|cmd)](https://github.com/Microsoft/vscode/tree/e29c517386fe6f3a40e2f0ff00effae4919406aa/resources)

The right command is moved at build time to `/Resources/app/bin/`


#### Path
How these scripts can be found by shell? `vscode` uses different mechanisms.

##### macOS 
At run time, when user executes the dedicated command in `vscode`, it installs `code` command in PATH. To do this, it makes a symlink: `/usr/local/bin/code -> /Applications/Visual Studio Code.app/Contents/Resources/app/bin/code` here: [src/vs/workbench/parts/cli/electron-browser/cli.contribution.ts](https://github.com/Microsoft/vscode/blob/e29c517386fe6f3a40e2f0ff00effae4919406aa/src/vs/workbench/parts/cli/electron-browser/cli.contribution.ts)

##### Windows
At installation time (InstallShield), `app/bin` is added to Path: [build/win32/code.iss#L785](https://github.com/Microsoft/vscode/blob/e29c517386fe6f3a40e2f0ff00effae4919406aa/build/win32/code.iss#L785)

##### Debian
At installation time, `app/bin` is symlinked to `/usr/bin/`: 
debian [resources/linux/debian/postinst.template#L8 ](https://github.com/Microsoft/vscode/blob/e29c517386fe6f3a40e2f0ff00effae4919406aa/resources/linux/debian/postinst.template#L8)
rpm [resources/linux/rpm/code.spec.template#L32](https://github.com/Microsoft/vscode/blob/e29c517386fe6f3a40e2f0ff00effae4919406aa/resources/linux/rpm/code.spec.template#L32)

### How to achieve that in Hyper?
#### CLI tool
We should the same as `vscode`: a javascript standalone tool, launched by a cmd/sh script.
#### Path
I didn't find how to modify user PATH in a postinstall step with `electron-builder`.
On macOS, we are distributing zipped app. We need to add symlink at runtime. We could do the same on Linux.
On Windows, we could maybe add the right registry key with the same way that we add Windows context menu entry: [app/index.js#L33](https://github.com/zeit/hyper/blob/d3a015466a87eb33138d0b6ab0e10b0cdffe8701/app/index.js#L33)

**Comment by chabou** [comment]
Due to [directory organization](https://github.com/Squirrel/Squirrel.Windows/blob/master/docs/using/install-process.md#install-location) on Windows, we should unlink old path and link new path after each update.

**Comment by albinekb** [comment]
Sounds like a plan @chabou 👍

**Comment by MAPESO** [comment]
@chabou I like it  👍

**Comment by leo** [comment]
Looks good to me! Thanks @chabou ❤️ 

But to be clear, the command will be called `hyper` and the official name for the binary (when referring to it in descriptions will be "Hyper CLI").

What I'm not sure on is whether or not we should also release the CLI on npm. But I don't think that makes any sense. Because even if people want to use it outside of Hyper, they can just do that, because it's in the `PATH`.

However, it might make sense when it comes to keeping it updated. I could imagine that some people might want to have a fresher version of Hyper CLI than the one their Hyper comes with. But all in all, I don't think that's a great idea since other versions of Hyper might support different configurations, etc. - so having a independently updated CLI doesn't sound good to me.

So feel free to add any additional opinions on that, but I think we should only have it inside Hyper (like @chabou suggested - no npm package).

**Comment by chabou** [comment]
@leo I'm totally agree with you and thank you for your clarification about `hyper` and "Hyper CLI".

As explained, we'll need to ship different boostrap script that depend on platform.
`vscode` use a gulp task to copy/rename files.

Until now, we have managed to not have a build script and to only depend on `electron-builder`. But I think that this time is over 😞 I didn't figure out how to copy/move some files, with `electron-builder` depending on targeted platform. Any idea?

If we need a build script:
Gulp do to much thing that we don't need, imo. What do you think about using [RunJS](https://github.com/pawelgalazka/runjs)? Minimalist, simple and javascript

**Comment by leo** [comment]
That's easy!

Options like [these](https://www.electron.build/configuration/contents) (look close, there are several different ones on this page - one of them will definitely work) allow us to copy files over to a directory of our choice in the bundle.

> May be specified in the platform options (e.g. in the mac).

I used this property in my [electron-next-skeleton](https://github.com/leo/electron-next-skeleton/blob/master/package.json#L14-L18) example project... 😊

**Comment by chabou** [comment]
OMG... How could I miss this 😱 ??!!

Thanks for your input @leo ❤️

**Comment by elijahgagne** [comment]
Any chance the Hyper CLI will let you create new tabs in Hyper that are started with arbitrary shells/arguments?

**Comment by albinekb** [comment]
That sounds like a great idea @elijahgagne, not in the first release but after that I don't see a reason not to. PR's welcome 🤗

**Comment by chabou** [comment]
Great idea yes.

With first release it will only be possible to create new tab with a specified path:
https://github.com/zeit/hyper/pull/2375

**Comment by chabou** [comment]
Fixed by #2375

**Comment by albinekb** [comment]
Awesome work with the CLI! This issue is resolved. Let's open new ones in the future
